Low level struct updates (#5602)
dotnet/csharplang: proposals/low-level-struct-improvements.md

=====

## Summary
This proposal is an aggregation of several different proposals for `struct` performance improvements: `ref` fields and the ability to override lifetime defaults. The goal being a design which takes into account the various proposals to create a single overarching feature set for low level `struct` improvements.

## Motivation
Earlier versions of C# added a number of low level performance features to the language: `ref` returns, `ref struct`, function pointers, etc. ... These enabled .NET developers to create write highly performant code while continuing to leverage the C# language rules for type and memory safety.  It also allowed the creation of fundamental performance types in the .NET libraries like `Span<T>`.

> As these features have gained traction in the .NET ecosystem developers, both internal and external, have been providing us with information on remaining friction points in the ecosystem. Places where they still need to drop to `unsafe` code to get their work, or require the runtime to special case types like `Span<T>`. 

Reviewer: ```suggestion
As these features have gained traction in the .NET ecosystem developers, both internal and external, have been providing us with information on remaining friction points in the ecosystem. Places where they still need to drop to `unsafe` code to get their work done, or require the runtime to special case types like `Span<T>`. 
```

> Today `Span<T>` is accomplished by using the `internal` type `ByReferenec<T>` which the runtime effectively treats as a `ref` field. This provides the benefit of `ref` fields but with the downside that the language provides no safety verification for it, as it does for other uses of `ref`. Further only dotnet/runtime can use this type as it's `internal`, so 3rd parties can not design their own primitives based on `ref` fields. Part of the [motivation for this work](https://github.com/dotnet/runtime/issues/32060) is to remove `ByReference<T>` and use proper `ref` fields in all code bases. 

Reviewer 2: ```suggestion
Today `Span<T>` is accomplished by using the `internal` type `ByReference<T>` which the runtime effectively treats as a `ref` field. This provides the benefit of `ref` fields but with the downside that the language provides no safety verification for it, as it does for other uses of `ref`. Further only dotnet/runtime can use this type as it's `internal`, so 3rd parties can not design their own primitives based on `ref` fields. Part of the [motivation for this work](https://github.com/dotnet/runtime/issues/32060) is to remove `ByReference<T>` and use proper `ref` fields in all code bases. 
```

This proposal plans to address these issues by building on top of our existing low level features. Specifically it aims to:

- Allow `ref struct` types to declare `ref` fields.

> - Allow the runtime to fully define `Span<T>` using the C# type system and remove special case type like `ByReference<T>`

Reviewer 2: are tehre any other speical types the runtime has that they could get rid of as well now? (just curious)

Author: No other special types but it's likely that once we have these features we can essentially eliminate the concept of restricted types. They are expressable now in terms of a `ref struct` with a `ref` field.

Reviewer 3: FWIW an existing special\restricted types being discussed for supporting https://github.com/dotnet/runtime/issues/45152 (`ref struct` + reflection with related "fast invoke" feature) is `TypedReference` which contains a loosely-typed `ByReference` expressed as a [`ByReference<byte>`](https://github.com/dotnet/runtime/blob/main/src/coreclr/System.Private.CoreLib/src/System/TypedReference.CoreCLR.cs#L16).

In order to support https://github.com/dotnet/runtime/issues/45152, a `static TypedReference FromRef<T>(ref T value)` [factory method](https://github.com/dotnet/runtime/compare/main...steveharter:FastInvoke2#diff-df560ce5bcf241f546b22cd124819fcf314e42ec480aaaf0c66fd653ce92e38cR72) is proposed which will create a `TypedReference` that references a `ref struct`. Today calling such a factory method is not possible with a `ref struct` since a `ref struct` type can't be used as a generic parameter, however that is a different issue; also note the method may be called "FromRefStruct" if necessary to distinguish between a `ref struct` and non-`ref-structs` so a new constraint like `where T : ref struct` can be added if necessary.

Author: > Today calling such a factory method is not possible with a ref struct since a ref struct type can't be used as a generic parameter, however that is a different issue; 

There is going to be a proposal soon-ish about how to make `ref struct` work with generic type arguments. I don't think we fully need that though to enable the `TypedReference` support. I think as a stop gap measure we could always just adjust the intrinsic `__makeref` to handle `ref struct`.

Consider the following method signature:

```c#
Span<T> CreateSpan<T>(ref T parameter)
{
    // The implementation of this method is irrelevant when considering the lifetime of the 
    // returned Span<T>. The rules disallow capture of `parameter` hence the return of 
    // CreateSpan<T> is always safe-to-escape to the calling method
}
```

By the existing span safety rules the return of an invocation of this method always has a //*safe-to-escape* value of *calling method*. That is because in the span safety document `ref` fields do not exist and hence there is no way for the return to capture `parameter` hence there is no constraint on the return.  The implementation of the method is completely irrelevant here, which is why it's omitted from the sample, because there is simply no way for the `ref` to be captured. This means no matter how this method is called the return is *safe-to-escape* to the *calling method*.

```c#
Span<T> CompatExample(ref int p)
{
    // Okay
    int local = 42;
    return CreateSpan<int>(ref local);

    // Okay
    Span<T> span = stackalloc int[1];

>     return CreateSpan<int>(ref span[0]);

Reviewer 2: wait... why is this ok? :)   furthermore... if it was legal before... woudl it be bad to become illegal?

Author: This was legal because there was no mechanism by which `CreateSpan(ref  int local)` could every capture and return the `ref`. Basically there is no way given the rules of C# and the safe APIs of dotnet/runtime to make the returned `Span<T>` refer to the passed in value. 

Yes you could do this using `MemoryMarshal` or `Unsafe` but those are APIs that are considered `unsafe`. They are essentially outside the bounds of span safety guidelines (and are documented as such)

Reviewer 4: @jaredpar, I think the "issue" here is that the code is confusing to read.

If you were to think about this from the perspective of the framework design guidelines and how the BCL names things; the "logical assumption" of `CreateSpan<int>(ref T)` is that it creates a `span` over the `ref` input and so the readers initial "assumption" is that `CreateSpan<int>(stackalloc)` is returning a span that captures the stack allocation.

The language prevents this today (outside of unsafe code), but its still not how the code reads. This is somewhat made more "confusing" by the fact that part of this doc is meant to enable scenarios where `CreateSpan(ref T)` capturing the input is possible.

Perhaps it would be more clear if this was `CreateSpanThatDoesntCaptureTheInput` (e.g. a method without `RefThisEscapes`) and `CreateSpanThatCapturesTheInput` (e.g. a method with `RefThisEscapes`)?
-- or better names, because names are hard and my suggestions are terrible 😄

Author: The goal of this section though is to really drive home the point that 

1. This method creates a `Span<T>` 
2. It can never, under any circumstances, capture that input value

If I name it say `CreateSpanThatDoesntCaptureTheInput` it might imply that this particular method does not capture the input but maybe some other method could. Where the problem is there is just no way this can happen, no matter the name or implementatoin.

Reviewer 4: > If I name it say CreateSpanThatDoesntCaptureTheInput it might imply that this particular method does not capture the input but maybe some other method could. Where the problem is there is just no way this can happen, no matter the name or implementatoin.

I thought that was the point of `RefThisEscapes`? `RefThisEscapes` means that you can have a span that captures the `ref input`. The language would then block it in the case of `stackalloc` because its unsafe.

That is, there are at least four basic scenarios to consider here, three of which are legal:
```csharp
// this is legal as `createdSpan` can't capture `ref span[0]` as it isn't attributed with `RefThisEscapes`
Span<T> stackSpan = stackalloc int[1];
var createdSpan = CreateSpanWithoutRefThisEscapes<int>(ref span[0]);

// this is legal as `createdSpan` can't capture `ref span[0]` as it isn't attributed with `RefThisEscapes`
Span<T> stackSpan = stackalloc int[1];
return CreateSpanWithoutRefThisEscapes<int>(ref span[0]);

// this is legal as even though `createdSpan` can capture `ref span[0]` since it is attributed with `RefThisEscapes`, it isn't escaping the scope of `stackSpan`
Span<T> stackSpan = stackalloc int[1];
var createdSpan = CreateSpanWithRefThisEscapes<int>(ref span[0]);

// this is ILLEGAL as `createdSpan` can capture `ref span[0]` and if it did, it would allow `stackSpan` to escape its scope
Span<T> stackSpan = stackalloc int[1];
return CreateSpanWithRefThisEscapes<int>(ref span[0]);
```

Author: > RefThisEscapes means that you can have a span that captures the ref input

No. The only thing that `RefThisEscapes` alters is that it makes it possible to return `this` by-ref from instance methods on `struct`. It does not change your ability to capture a `ref` as a field.

Reviewer 4: Maybe I'm getting the attributes mixed up then?

This allows ref fields, which means that you can functionally `new Span(ref param)` and the created `Span` will have the scope of the passed input.

So if the `CreateSpan` call uses that method and returns the span, its lifetime is the same as the input and the compiler does the right analysis.

    // Okay
    int[] array = new int[42];
    return CreatSpan(ref array[0]);
}
```

> All of the above `return` statements are legal by our existing rules because the return of `CreateSpan` is always *safe-to-escape* to the *calling method*. The challenge in this proposal, which comes up many times, is they **must** remain legal in the version of the language which implements these features and / or when `Span<T>` moves to using `ref` fields. The above patterns can legally exist today and cannot become errors when moving to a new language version. 

Reviewer 2: > and cannot become errors when moving to a new language version. 

Would it be ok for them to become errors if there was true memory unsafety before?

Reviewer 4: Warning wave maybe?

There is, on the BCL side, the interesting case of `MemoryMarshal.CreateSpan()` which functionally works like the proposed constructor (and via an internal `Span(ref T ptr, int length)` constructor):
```csharp
[RefFieldEscapes]
public Span(ref T value) { }
```

The BCL will need to decide how it works moving forward, including whether it starts erroring just like the proposed constructor, or continues allowing users to do "unsafe" things (potentially with an analyzer, etc).

Author: > Would it be ok for them to become errors if there was true memory unsafety before?

This pattern is completely safe though. The only way this could be unsafe is if `Span<T>` had a constructor in the form of `Span(ref T value)` which is explicitly disallowed by the span safety rules. Yes `Span<T>` does have this CTOR but it's `internal` and known to be `unsafe`. 

So there is nothing wrong with these patterns. They are safe, legal and honestly expected. The issue is that we must maintain those properties going forward.

Author: > Warning wave maybe?

How would a warning wave help though? Consider that by adding `[RefFieldEscapes]` to a method you change the semantics from the perspective of the caller. Or consider it this way:

```
Span<int> MyMethod() {
  int local = 42;
  Span<int> span = CreateSpan(ref local); 
  return span;
}
``` 

This is 100% legal today. If we suddenly warn customers into using `[RefFieldsEscape]` on such methods then it suddenly breaks callers of that method.

Reviewer 2: Just so i understand with that example.  You're sayign that that code coudl exist and have legit uses (for example, not captuing 'local' within the returned span).  And as such, because of hte legal reasons, we couldn't jsut always start warning/error'ing here.

*however*, what we can do is have CreateSpan announce it's intentions about what happens with the passed in 'ref'.  If it ignored it (as an example) it would just do nothing, and the above would continue to be fine.  However, if it ends up pushing that parameter into the span it returns, then it shoudl have this attribute (virally spread hopefully from `Span<T>` inside it) so that hte caller can now error *in that case*.

--

In other words, the attribute allows cooperation between the caller and callee here on legal cases that should continue working, vs bad cases that ideally should break because they're entire memory unsafe.  Is that right?

Reviewer 4: > This is 100% legal today.

Right, and in the case where `CreateSpan` isn't updated to use some method/constructor marked `RefFieldsEscape` to create the returned span, it continues being "safe".
- If `CreateSpan` were updated to use `[RefFieldEscapes] public Span(ref T value) { }`, then I would assume the compiler errors as it expects `CreateSpan` to then be equivalently annotated, etc. This would be a breaking change for a library to make and therefore not recommended.

There are, however, scenarios like with `MemoryMarshal.CreateSpan` where its known to be "unsafe" and where the compiler (or an analyzer) could surface a warning that it functionally works like `[RefFieldEscapes] public Span(ref T value) { }` which would help catch problematic cases (particularly for users transitioning from "old unsafe logic" to "new safe logic").

You could also consider cases like `return new Span(&local, 1)` where its unsafe and legal, but its also a detectable problem and where a warning being surfaced is probably a "good idea" (maybe these are better suited to a library analyzer though)

Author: > Just so i understand with that example. You're sayign that that code coudl exist and have legit uses (for example, not captuing 'local' within the returned span). And as such, because of hte legal reasons, we couldn't jsut always start warning/error'ing here.

Exactly. It's perfectly okay to create such an API and give it legitamate uses. That exact pattern exists a few places in dotnet/runtime today (`SecureString` and some parts of `DateTime` parsing I was able to find). 

> In other words, the attribute allows cooperation between the caller and callee here on legal cases that should continue working, vs bad cases that ideally should break because they're entire memory unsafe. Is that right?

yep.

Author: > There are, however, scenarios like with MemoryMarshal.CreateSpan where its known to be "unsafe" and where the compiler (or an analyzer) could surface a warning ... 

I'd actually vote in `net6.0` that we obsolete this API. Basically say "use `new Span<T>`" instead please. That would force the caller to use `unsafe` in the cases where actual `unsafe` behavior was still needed. 

I'm not sure how feasible that is though.

Reviewer 2: > yep.

Gracias!  you probably did write all that out in the PR.  but it helped to get that clarity here :)

Further the design must ensure that once `ref` fields exist that implementations of `CreateSpan<T>` cannot suddenly begin capture the input arguments by `ref`. For example: 

```c#
Span<T> CreateSpan<T>(ref T parameter)
{
    // Error: if this works it breaks the above compat requirements because the implementation
    // is now capturing `ref` state.

>     return Span<T>(ref parameter);

Reviewer 2: i'm not sure if this is a great example.  since the ref is captured, but only in somethign passed back to the caller, this is safe right (and i'm not sure how it breaks a compat requirement "if this works").

Author: Let me try expanding that sample out a bit. Essentially getting through the parts that will work (calling the ctor) and the parts that won't (returning it).

The set of changes to our span safety rules necessary to allow `ref` fields is small and targeted. However this section of the proposal is quite involved. The reason is to make sure the reader understands both the new rules and **why** they were chosen. There are a number of subtle interactions between `ref` fields and our [compat considerations](#compat-considerations) that are easy to miss. This section is meant to call these out and explain how they fit into the chosen rules.

The [new Span<T> definition](#new-span) also reveals several challenges that must be resolved for the lifetime of types that contain `ref` fields. They must both take into account the lifetime of captured values as well as the compat considerations.

```c#
Span<int> CreatingAndReturningSpan(ref int parameter)
{
    int local = 42;

    // This must be an error because the created Span<T> is storing a reference to the 
    // stack. It cannot be safely returned.
    return new Span<int>(ref local);

    // This must be an error because of the compat issues of the existing span safety 
    // rules.

>     return new Span<int>(ref parameter);

Reviewer 2: i'm not really getting this statement (i feel like it hasn't been explained well enough).  Why is it necessary for something safe to be an error.  compat issues tell me that prior working code should continue to work.  but it doesn't tell my why something must be an error.

Author: I put a lot of the details into the earlier section but didn't have the full set of examples. I'm just going to expand the earlier section to include these, and link to that from here.

<a name="ref-fields-escape"></a>

To accomplish this the span safety rules will be changed as follows. First for the constructor of a `ref struct` that directly contains a `ref field`: 
- If the constructor is annotated with `[RefFieldEscapes]` then `ref` fields can be initialized with any value that has *ref-safe-to-escape* to the *calling method*
- Else `ref` fields can only be initialized with known heap values.

These rules ensure that the caller can understand when constructors will or will not capture parameters by `ref`. 

Next the rules for method invocation will change as follows when the target method is annotated with `[RefFieldEscapes]`:
- If the current method is not annotated with `[RefFieldEscapes]` then
    - If all arguments have a *ref-safe-to-escape* scope of *heap* then the *safe-to-escape* scope of the return is to the *calling method*
    - Else the *safe-to-escape* scope is to the *current method*
- Else the *safe-to-escape* scope of the return is the minimum of 
    - The existing *safe-to-escape* calculation for method invocation

>     - All of the *ref-safe-to-escape* values of `ref` and `in` arguments

Reviewer 2: whooboy.  this last part went entirely over my head :)

Author: That language takes some getting used to but it's how the original span safety doc was written. I want this doc to provide a map for how these features will fit into our existing spec. Essentially what changes do we need to make to extend our existing rules.

> The design of `[RefFieldEscape]` will be discussed in detail [later in the propsal](#reffieldescapes).

Reviewer 2: ```suggestion
The design of `[RefFieldEscape]` will be discussed in detail [later in the proposal](#reffieldescapes).
```

        // all of our existing rules
        return new RS(new int[1]);
    }

    // This method demonstrates that calling CreateRS2 has the same restrictions as RS(ref int) 
    // constructor. This demonstrates that factory methods are as flexible as constructors 
    // concerning `ref` field initialization
    [RefFieldEscapes] 
    static RS CreateRS3(ref int parameter)
    {
        // Okay: This method has [RefFieldEscapes] hence the safe-to-escape of `rs` is the minimim
        // of the ref-safe-to-escape of the arguments. In this case that is `parameter` which is 
        // to the calling method
        RS rs = new CreateRS2(ref parameter);
        return rs;

        int local = 42;

        // Error: In this case the ref-safe-to-escape of `local` is to the current method hence 
        // it is also the safe-to-escape of the method invocation.
        return new CreateRS2(ref local);
    }
}
```

The samples here have the same patterns as the [compat considerations](#compat-considerations) above. This means it will allow the introduction of `ref` fields without breaking existing code.

> Constructor chaining needs to consider these new rules as well. When the original constructor calls a chained constructor via `:this(...)` the chained constructor effectively escapes from the original. That means a chained constructor call is olny legal if the *safe-to-escape* value is not smaller than the original constructor one. This will be accomplished with the following rules:

Reviewer 2: ```suggestion
Constructor chaining needs to consider these new rules as well. When the original constructor calls a chained constructor via `:this(...)` the chained constructor effectively escapes from the original. That means a chained constructor call is only legal if the *safe-to-escape* value is not smaller than the original constructor one. This will be accomplished with the following rules:
```

The design of `[RefFieldEscape]` will be discussed in detail [later in the proposal](#reffieldescapes).

Let's examine these rules in the context of samples to better understand their impact and how they maintain the required compat considerations.

```c#
ref struct RS
{
    ref int _field;

    public RS(int[] array, int index)
    {
        // Okay: even though there is no [RefFieldEscapes] annotation the fields are initialized
        // with known heap values
        _field = ref array[index];
    }

    [RefFieldEscapes]
    public RS(ref int i)
    {
        // Okay: the constructor is annotated with [RefFieldEscapes] which allows for fields to be
        // initialized with any value that is ref-safe-to-escape to calling method. That is the 
        // case for `i` hence this is legal
        _field = ref i;
    }

    public RS(ref (int, int) tuple)
    {
        // Error: even though the ref-safe-to-escape of `tuple.Item1` is to the calling method, the 
        // constructor is not annotated with [RefFieldEscapes] hence only known heap values are 
        // allowed
        _field = ref tuple.Item1;
    }

    static RS CreateRS(ref int parameter)
    {

>         // Error: The RS(ref int) ctor is annotated as [RefFieldEscapes] however CreateRS is 

Author: Generics should not add any particular complication here. This problem only comes into effect when: 

- A `ref struct` with a `ref` field is returned from a method. Either by the method return or via a `out` / `ref` parameter
- There is an additional `ref`, `out` or `in` parameter of any type

Hence generics don't impact this much at all because `ref struct` can't be generic. Even if we add support for generic `ref struct` then it's an opt in scenario, like `where T : ref struct`, hence it should still be very limited. 

Note: in searching through the BCL today I've only found a very limited number of these methods and most are non-public. If you are aware of more I'd like to hear about them.

}
```

The rules for `ref` assignment also need to be adjusted to account for `ref` fields.  This design only allows for `ref` assignment of a `ref` field during object construction or when the value is known to refer to the heap. Object construction includes in the constructor of the declaring type, inside `init` accessors and inside object initializer expressions. Further the `ref` being assigned to the `ref` field in this case must have *ref-safe-to-escape* greater than the receiver of the field: 

- Constructors: The value must be *ref-safe-to-escape* outside the constructor
- `init` accessors:  The value limited to values that are known to refer to the heap as accessors can't have `ref` parameters
- object initializers: The value can have any *ref-safe-to-escape* value as this will feed into the calculation of the *safe-to-escape* of the constructed object by existing rules.

> A `ref` field can only be `ref` assigned outside a constructor when the value is known to refer to the heap. That is allowed because it is both safe at the assignment location (meets the field assignment rules for ensuring the value being assigned has a lifetime at least as large as the receiver) as well as requires no updates to the existing method invocation rules. 

Reviewer: >when the value is known to refer to the heap [](http://example.com/codeflow?start=63&length=44)

What if the lifetime of the value is known to encompass the lifetime of the `ref struct` containing the field?

Author: True it could be expressed that way. A `ref struct` really only has two lifetime states: usable within the current method, returnable to the containing method. If we make it such that the field lifetime has to be at least the containing `ref struct` then that would also allow for a bit more flexibility. 

I need to think about this a bit though and make sure I'm not missing a subtle rule here.

- `init` accessors:  The value limited to values that are known to refer to the heap as accessors can't have `ref` parameters
- object initializers: The value can have any *ref-safe-to-escape* value as this will feed into the calculation of the *safe-to-escape* of the constructed object by existing rules.

> A `ref` field can only be `ref` assigned outside a constructor when the value is known to have a lifetime greater than the reciever. Specifically: 

Reviewer: Perhaps update to allow a lifetime that is the same span:
```suggestion
A `ref` field can only be `ref` assigned outside a constructor when the value is known to have a lifetime greater than or equal to the receiver. Specifically: 
```

}
```

Detailed Notes:
- A `ref` field can only be declared inside of a `ref struct` 
- A `ref` field cannot be declared `static`
- A `ref` field can only be `ref` assigned 
    - in the constructor of the declaring type
    - when the RHS is known to be a heap location
- The reference assembly generation process must preserve the presence of a `ref` field inside a `ref struct` 
- A `readonly ref struct` must declare its `ref` fields as `readonly ref`
- The span safety rules for constructors, fields and assignment must be updated as outlined in this document.
- The span safety rules need to include the definition of `ref` values that "refer to the heap". 

> ### Provide lifetime annotations

Reviewer 5: I wonder if we should have another annotation for extremely specialized cases where the method implementor wants to prohibit escape, but its not a capturing escape. For instance, the StackAlloc api I proposed in a hackathon a few months ago.
```
class RuntimeHelpers
{
    static Span<T> StackAlloc<T>(int count) {... }
}
```

Ideally there would be an annotation that allows specifying that the Span<T> returned from StackAlloc is not safe to escape.

OTOH, this may be so incredibly niche that we should just special case the function and move on.

Author: I thought about this too. I think it would come down to how many of these functions we ended up creating. If it's only going ot be a small number, say < 5, and they're all effectively runtime intrinsics then possibly we just special case them. If it's more than that though then I'd favor adding an annotation for them.

While the defaults have allowed broad adoption of `ref struct` within .NET they do create a number of friction points in low level code. For example the inability to return fields of `struct` by `ref` from instance methods. These friction points often force developers to resort to using `unsafe` which de-values our `ref struct` efforts. 

The lifetime defaults for these locations is not fundamental to correctness. For example parameters could default to *ref-safe-to-escape* to within the current method or `this` in a `struct` could be *ref-safe-to-escape* to the calling method in our span safety rules and it would not make `ref struct` usage unsafe. It would simply impact the usability of the resulting feature.

This, and several other friction points, can be removed if the language provides developers a way to invert the defaults by applying attributes to specific locations. The language can recognize these attributes and simply adjust the lifetime calculation for locations when evaluating span safety.

#### RefThisEscapes

> One of the most notable friction points in the inability to return fields by `ref` in instance members of a `struct`. This means developers can't create `ref` returning methods / properties and have to resort to exposing fields directly. This reduces the usefulness of `ref` returns in `struct` where it is often the most desired. 

Reviewer 6: ```suggestion
One of the most notable friction points is the inability to return fields by `ref` in instance members of a `struct`. This means developers can't create `ref` returning methods / properties and have to resort to exposing fields directly. This reduces the usefulness of `ref` returns in `struct` where it is often the most desired. 
```

```c#
struct S
{

>      _field;

Reviewer: ```suggestion
     int _field;
```

    // Error: this, and hence _field, can't return by ref
    public ref int Prop => ref _field;
}
```

The [rationale](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md#struct-this-escape) for this default is reasonable but there is nothing inherently wrong with a `struct` escaping `this` by reference, it is simply the default chosen by the span safety rules. 

To fix this the attribute `System.Runtime.CompilerServices.RefThisEscapesAttribute` will be used to specify that `this` is *ref-safe-to-escape* to the calling method. It can be applied to individual methods or properties in `struct`. When applied to `struct` declaration itself it is treated as if it were applied to all methods / properties in the type. 

```c#
// Option 1 
struct S
{
    int _field;

    // The ref-safe-to-escape for this is now the calling method hence this is legal
    [RefThisEscapes]
    public ref int Prop => ref _field;
}

// Option 2 
// The ref-safe-to-escape for this is now the calling method on all members

> [ThisRefEcapes]

Reviewer: ```suggestion
[ThisRefEscapes]
```

struct S
{
    int _field;

    public ref int Prop => ref _field;
}
```

This will naturally, by the existing rules in the span safety spec, allow for returning transitive fields in addition to direct fields.

```c#
[RefThisEscapes]
struct Child
{
    int _value;
    public ref int Value => ref _value;
}

[RefThisEscapes] 
struct Container
{
    Child _child;

    // In this case the ref-safe-to-escape of `_child` is to the calling method because that is 
    // the value of `this` and fields derive it from their receiver. From there method invocation 
    // rules take over 
    public ref int Value => ref _child.Value;
}
```

This require the following changes to the span safety document: 

- The method invocation rules will include the *ref-safe-to-escape* value of the receiver when it is marked by `[RefThisEscapes]`
- The parameter lifetime rules will change to note that the *ref-safe-to-escape* of `this` is dependent on `[RefThisEscapes]`

Detailed Notes:
- An instance method or property annotated with `[RefThisEscapes]` has *ref-safe-to-escape* of `this` set to the *calling method*

> - A `struct` annotated with `[RefThisEscapes]` has the same effect of annotating every instance method and property with `[RefThisEscapes]`

Reviewer 6: I think it would be helpful for an example that demonstrates why one may want to annotate a single method/property as opposed to the entire type.

Reviewer 4: "Anonymous structs/unions" used in interop.

There are many cases where `FieldOffset` is not viable because some other member in the struct is variable sized (`nint`, `nuint`, a `pointer`, etc). In those cases, the only solution is to have something like:
```csharp
public unsafe partial struct MyStruct
{
    public _Anonymous_e__Union Anonymous;
    
    public unsafe partial struct _Anonymous_e__Union
    {
        public float m11;
    }
} 
```

In such a scenario, there are a lot of "usability" issues with forcing users to do `value.Anonymous.m11` rather than just allowing `value.m11`. The "simplest" solution here is to expose a `ref property`:
```csharp
public unsafe partial struct MyStruct
{
    public _Anonymous_e__Union Anonymous;
    
    public ref float m11 => ref MemoryMarshal.GetReference(MemoryMarshal.CreateSpan(ref Anonymous.m11, 1));
    
    public unsafe partial struct _Anonymous_e__Union
    {
        public float m11;
    }
} 
```

However, this is unsafe and doesn't properly track the lifetime of the `ref`. With the introduction of `RefThisEscapes`, this becomes:
```csharp
public unsafe partial struct MyStruct
{
    public _Anonymous_e__Union Anonymous;
    
    [RefThisEscapes]
    public ref float m11 => ref Anonymous.m11;
    
    public unsafe partial struct _Anonymous_e__Union
    {
        public float m11;
    }
} 
```

You wouldn't want to annotate the entire type, because things like `operator ==(in MyStruct left, in MyStruct right)` take a ref, but don't return a reference to the `this` field.

Reviewer 6: Thanks. Great example.

The semantics of this attribute, and how it impacts span safety rules, are described [above](#ref-fields-escape)

#### DoesNotEscape

> One source of repeated friction in low level code is the default escape for parameters is permissive. They are *safe-to-escape* to the *calling method*. This is a sensible defaults because it lines up with the coding patterns of .NET as a whole. In low level code though there is a larger use of  and `ref struct` and this default can cause friction with other parts of the span safety rules.

Reviewer: ```suggestion
One source of repeated friction in low level code is the default escape for parameters is permissive. They are *safe-to-escape* to the *calling method*. This is a sensible default because it lines up with the coding patterns of .NET as a whole. In low level code though there is a larger use of  `ref struct` and this default can cause friction with other parts of the span safety rules.
```

The other place where parameter default escape scope causes friction is when they are re-assigned within a method body. For instance if a method body decides to conditionally apply escaping to input by using stack allocated values. Once again this runs into some friction.

```c#
void WriteData(ReadOnlySpan<char> data)
{
    if (data.Contains(':'))
    {
        Span<char> buffer = stackalloc char[256];
        Escape(data, buffer, out var length);

        // Error: Cannot assign `buffer` to `data` here as the safe-to-escape

>         // scope of `buffer` is to the current method scope while `buffer` is

Reviewer 6: ```suggestion
        // scope of `buffer` is to the current method scope while `data` is
```

To remove this friction the language will provide the attribute `System.Runtime.CompilerServices.DoesNotEscapeAttribute`. This can be applied to parameters or instance methods of a `ref struct`. When applied to parameters the *safe-to-escape* and *ref-safe-to-escape* scope will be to the *current method*. When applied to instance members of a `ref struct` the same limitations will apply to the `this` parameter

```c#
class C
{
    static Span<int> M1(Span<int> p1, [DoesNotEscape] Span<int> p2)
    {
        // Okay: the safe-to-escape here is still outside the enclosing scope
        // of the current method.
        return p1; 

        // Error: the [DoesNotEscape] attribute changes the safe-to-escape*
        // to be limited to the current method scope. Hence it cannot be 
        // returned
        return p2; 

        // Error: `local` has the same safe-to-escape as `p2` hence it cannot
        // be returned.
        Span<int> local = p2;

>         return p2; 

Reviewer 6: ```suggestion
        return local; 
```

}
```

This attribute cannot be used on `ref` or `out` parameters. Those always implicitly escape to the calling method (that is how `ref` works). Instead `in` is likely a more appropriate designation for those scenarios.

To account for this change the "Parameters" section of the span safety document will be updated to include the following bullet:

> - If the parameter is marked with `[DoesNotEscape]`it is *safe-to-escape* is to the *current method*

Reviewer: ```suggestion
- If the parameter is marked with `[DoesNotEscape]` it is *safe-to-escape* to the *current method*
```

        // of the current method and hence the same as the receiver.
        _field = p1;

        // Error: the [DoesNotEscape] attribute changes the *safe-to-escape* 
        // to be limited to the current method scope. Hence it cannot be 

>         // assigned to a receiver than has a *safe-to-escape* scope outside the 

Reviewer 6: ```suggestion
        // assigned to a receiver that has a *safe-to-escape* scope outside the 
```

Several ideas for having implicit opt-in to `ref` capture were explored and discarded: 

- Special casing constructors. It is possible to have constructors of `ref struct` that **directly** define `ref` fields be implicitly opt-in to `[RefFieldEscapes]` semantics. However this does not generalize to factory methods and hence is not a general solution that we can use.
- Special casing methods that return `ref struct` that define a `ref`. There are no such methods today because `ref` fields do not exist hence we could say that methods which return `ref struct` that defined a `ref` field have opted-in to `[ReFFieldsEscape]` semantics. This works but it essentially prevents any existing `ref struct` from adding `ref` fields. Doing so would cause span safety rules to be interpreted differently in all methods that returned the type. 

These implicit opt-in strategies all have significant holes while an explicit opt-in is fully generalizable and makes the span safety rule different explicit in the code.

### Why isn't assignability to ref field a scope?

> It may seem attractive to think of the ability to assign to a `ref` field as just another scope concept instead of an orthogonal concept. Essentially *assignable to field* as an escape scope between *heap* and *calling method*. This falls a part though when looking at practical examples. Consider for example the following: 

Reviewer: ```suggestion
It may seem attractive to think of the ability to assign to a `ref` field as just another scope concept instead of an orthogonal concept. Essentially *assignable to field* as an escape scope between *heap* and *calling method*. This falls apart though when looking at practical examples. Consider for example the following: 
```

```c#

> void M(ref int parameter)

Reviewer: >int parameter [](http://example.com/codeflow?start=11&length=13)

Should `parameter` be used below?

Author: It just shouldn't be there at all

{
    int local = 0;
    Span<T> span = new(ref local);

> }

Reviewer: Assuming the constructor has `[RefFieldsEscape]`, doesn't this fit the escape scope model since `local` and `span` both have the same scope?

Author: Correct. If `[RefFieldEscapes]` is on the ctor then both `local` and `span` have the same *safe-to-escape* scope

```

The design explicitly wants to ensure this is legal. Yet the *ref-safe-to-escape* scope of `local` is *current method* and it can't be anything greater. At the same time it must be assignable to a `ref` field. Further we need to ensure that `span` is itself assignable to a `ref` field of a different type and it also has *current method* as the *ref-safe-to-escape* scope.

The only practical way to represent whether a location can assign to a `ref` field is as a separate concept from escape scopes.

## Open Issues

### Take the breaking change

> Consider for a minute a design where the compat problem was approached from the other direction. Effectively make it such that every method was implicitly `[RefFieldsEscape]` and then have an attribute that restores the span safety rules in place today. Say `[RefFieldDoesNotEsacpe]`. 

Reviewer: ```suggestion
Consider for a minute a design where the compat problem was approached from the other direction. Effectively make it such that every method was implicitly `[RefFieldsEscape]` and then have an attribute that restores the span safety rules in place today. Say `[RefFieldDoesNotEscape]`. 
```

> This would be a breaking change and it's easy to construct code samples that trigger an error when upgrading to a new version of C# (as demonstrated [here](#new-span-challenges)). It's hard to determine though how prevelant these types of patterns are. Essentially methods which have both the following attributes:

Reviewer: ```suggestion
This would be a breaking change and it's easy to construct code samples that trigger an error when upgrading to a new version of C# (as demonstrated [here](#new-span-challenges)). It's hard to determine though how prevalant these types of patterns are. Essentially methods which have both the following attributes:
```

- Have a `Span<T>` or `ref struct` which has a `Span<T>` as
    - The return type 
    - A `ref` or `out` parameter
- Take a `ref`, `in` or `out` parameter 

If this is very low then it's possible that a breaking change could be acceptable here. 

The challenge though is the breaking change and the fix for the breaking change are possible in different libraries. Consider the following:

```c#
// Widget.Library.dll
Span<int> CreateSpan(ref int i)
{
    ...
    return new Span<int>(new int[i]);
}

// App.exe
Span<int> Method()
{
    int local = 42;

>     var span = new Span<int>(ref local);

Reviewer: ```suggestion
    var span = CreateSpan(ref local);
```

- Special casing methods that return `ref struct` that define a `ref`. There are no such methods today because `ref` fields do not exist hence we could say that methods which return `ref struct` that defined a `ref` field have opted-in to `[ReFFieldsEscape]` semantics. This works but it essentially prevents any existing `ref struct` from adding `ref` fields. Doing so would cause span safety rules to be interpreted differently in all methods that returned the type. 

These implicit opt-in strategies all have significant holes while an explicit opt-in is fully generalizable and makes the span safety rule different explicit in the code.

### Reference Assemblies
A reference assembly for a compilation using features described in this proposal must maintain the elements that convey span safety information. That means all lifetime annotation attributes and `[RefFieldEscapes]` must be preserved in their original position. Any attempt to replace or omit them can lead to invalid reference assemblies.

> Representing `ref` fields is more nuanaced. Ideally a `ref` field would appear in a reference assembly as would any other field. However a `ref` field represents a change to the metadata format and that can cause issues with tool chains that are not updated to understand this metadata change. A concrete example is C++/CLI which will likely error if it consumes a `ref` field. Hence it's advantageous if `ref` fields can be omitted from reference assemblies in our core libraries. 

Reviewer: ```suggestion
Representing `ref` fields is more nuanced. Ideally a `ref` field would appear in a reference assembly as would any other field. However a `ref` field represents a change to the metadata format and that can cause issues with tool chains that are not updated to understand this metadata change. A concrete example is C++/CLI which will likely error if it consumes a `ref` field. Hence it's advantageous if `ref` fields can be omitted from reference assemblies in our core libraries. 
```

> A `ref` field by itself has no impact on span safety rules. As a conecrete example consider that flipping the existing `Span<T>` defintion to use a `ref` field has no impact on consumption. Hence the `ref` itself can be omitted safely. However a `ref` field does have other impacts to consumption that must be preserved: 

Reviewer: ```suggestion
A `ref` field by itself has no impact on span safety rules. As a concrete example consider that flipping the existing `Span<T>` definition to use a `ref` field has no impact on consumption. Hence the `ref` itself can be omitted safely. However a `ref` field does have other impacts to consumption that must be preserved: 
```

- A `ref struct` which has a `ref` field is never considered `unmanaged` 
- The type of the `ref` field impacts infinite generic expansion rules. Hence if the type of a `ref` field contains a type parameter that must be preserved 

Given those rules here is a valid reference assembly transformation for a `ref struct`: 

```c#
// Impl assembly 
ref struct S<T> {
    ref T _field;
}

// Ref assembly 
ref struct S<T> {

>     object _o; // force unmanaged 

Reviewer: Doesn't the additional field change the size of the `ref struct` for the consumer?

Author: Generally size is calculated at runtime based on the impl assembly, not the reference assembly. It's generally considered acceptable for reference assemblies to add / remove fields. Been a common practice in GenAPI for a number of years.

}
```

Given that parameters are restricted in this way we will also update the "Method Invocation" section to relax its rules. In all cases where it is considering the *safe-to-escape* lifetimes of arguments the spec will change to ignore those arguments which line up to parameters which are marked as `[DoesNotEscape]`. Because these arguments cannot escape their lifetime does not need to be considered when considering the lifetime of returned values.

For example the last line of calculating *safe-to-escape* of returns will change to 

> the safe-to-escape of all argument expressions including the receiver. **This will exclude all arguments that line up with parameters marked as [DoesNotEscape]**

Detailed Notes:
- The `[DoesNotEscape]` and `[RefThisEscapes]` cannot be combined on the same method 

> - The `[DoesNotEsacpe]` attribute cannot be used on parameters that are `ref`, `out` or `in`.

Reviewer: ```suggestion
- The `[DoesNotEscape]` attribute cannot be used on parameters that are `ref`, `out` or `in`.
```

> the safe-to-escape of all argument expressions including the receiver. **This will exclude all arguments that line up with parameters marked as [DoesNotEscape]**

Detailed Notes:

> - The `[DoesNotEscape]` and `[RefThisEscapes]` cannot be combined on the same method 

Reviewer 6: Isn't `DoesNotEscape` the default for member functions of a `ref struct`? Trying to confirm my understanding here. If so, can we should have an example where one would want to have `DoesNotEscape` applied to a member function.

Author: For a instance member of a `struct` of any kind the `this` parameter is *safe-to-escape* to the calling method but *ref-safe-to-escape* within the current method. The `[DoesNotEscape]` annotation moves both to current method. 

An example where you would need it is the following: 

```c# 
ref struct MyType { 
  Span<int> _span;
  public MyType(Span<int> span) { }
  public void Method(ref ValueStringBuilder builder) { ... }
}
```

This method falls squarely into the [method arguments must match](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-7.2/span-safety.md#method-arguments-must-match) problem because both `this` and `builder` are *safe-to-escape* to the calling method. That means from the point of view of the compiler `builder` could end up smuggling out `ref` state from `this`. Yes realistically this can't happen do to no common data but compiler doesn't do field level analysis, it's just looking at the escape scopes of the locations.

This mean though practically that you run into problems like this: 

``` 
void Example(ref ValueStringBuilder builder) {
  var myType = new MyType(stackalloc int[42]); 
  
 // Error: builder and myType have different lifetimes
  myType.Method(ref builder); 
``` 

This can be fixed though by noting that `this` does not escape back into the parameters 

``` 
[DoesNotEscape]
public void Method(ref ValueStringBuilder builder) { ... }
```

### Safe fixed size buffers
The language will relax the restrictions on fixed sized arrays such that they can be declared in safe code and the element type can be managed or unmanaged.  This will make types like the following legal:

```c#
internal struct CharBuffer
{
    internal char Data[128];
}
```

These declarations, much like their `unsafe` counter parts, will define a sequence of `N` elements in the containing type. These members can be accessed with an indexer and can also be converted to `Span<T>` and `ReadOnlySpan<T>` instances.

When indexing into a `fixed` buffer of type `T` the `readonly` state of the container must be taken into account.  If the container is `readonly` then the indexer returns `ref readonly T` else it returns `ref T`. 

> Accessing a `fixed` buffer without an indexer has no natural type however it is convertible to `Span<T>` types. In the case the container is `readonly` the buffer is implicitly convertible to `ReadOnlySpan<T>`, else it can implicitly convert to `Span<T>` or `ReadOnlySpan<T>` (the `Span<T>` conversion is considered *better*). 

Reviewer 6: > the `Span<T>` conversion is considered *better*

Why is this true?

Reviewer 4: I would assume because the data is mutable so `Span<T>` is a better fit.

Plus, if you get a `Span<T>`, you can explicitly convert to `ROSpan<T>`. You cannot trivially do the inverse (and the inverse is generally considered "unsafe")

Author: Indeed that is the reason.

The resulting `Span<T>` instance will have a length equal to the size declared on the `fixed` buffer. The *safe-to-escape* scope of the returned value will be equal to the *safe-to-escape* scope of the container, just as it would if the backing data was accessed as a field.

For each `fixed` declaration in a type where the element type is `T` the language will generate a corresponding `get` only indexer method whose return type is `ref T`. The indexer will be annotated with the `[RefThisEscapes]` attribute as the implementation will be returning fields of the declaring type. The accessibility of the member will match the accessibility on the `fixed` field.

For example, the signature of the indexer for `CharBuffer.Data` will be the following:

```c#
[RefThisEscapes]

> internal ref char <>DataIndexer(int index) => ...;

Reviewer 6: Is the `<>` correct in the above signature?

Author: Yes it's meant to indicate that it's a compiler generated member that can't be directly accessed by consumer code. The `<>` is part of all synthesized non-acecssible members we generate (usually has a bit of text in between)

```

If the provided index is outside the declared bounds of the `fixed` array then an `IndexOutOfRangeException` will be thrown. In the case a constant value is provided then it will be replaced with a direct reference to the appropriate element. Unless the constant is outside the declared bounds in which case a compile time error would occur.

There will also be a named accessor generated for each `fixed` buffer that provides by value `get` and `set` operations. Having this means that `fixed` buffers will more closely resemble existing array semantics by having a `ref` accessor as well as byval `get` and `set` operations. This means compilers will have the same flexibility when emitting code consuming `fixed` buffers as they do when consuming arrays. This should be operations like `await` over `fixed` buffers easier to emit. 

This also has the added benefit that it will make `fixed` buffers easier to consume from other languages. Named indexers is a feature that has existed since the 1.0 release of .NET. Even languages which cannot directly emit a named indexer can generally consume them (C# is actually a good example of this).

> There will also be a by value `get` and `set` accessor generated for every 

Reviewer 6: Incomplete thought?

Author: The thought was complete, the text was lacking 😄

> The backing storage for the buffer will be generated using the `[InlineArray]` attribute. This is a mechanism discussed in [isuse 12320](https://github.com/dotnet/runtime/issues/12320) which allows specifically for the case of efficiently declaring sequence of fields of the same type. This particular issue is still under active discussion and the expectation is that the implementation of this feature will follow however that discussion goes.

Reviewer 6: ```suggestion
The backing storage for the buffer will be generated using the `[InlineArray]` attribute. This is a mechanism discussed in [issue 12320](https://github.com/dotnet/runtime/issues/12320) which allows specifically for the case of efficiently declaring sequence of fields of the same type. This particular issue is still under active discussion and the expectation is that the implementation of this feature will follow however that discussion goes.
```

## Considerations

### Keywords vs. attributes
This design calls for using attributes to annotate the new lifetime rules. This also could've been done just as easily with contextual keywords. For instance: `scoped` and `escapes` could have been used instead of `DoesNotEscape` and `RefThisEscapes`.

> Keywords, even the contextual ones, have a much heavier weight in the language than attributes. The use cases these features solve, while very valuable, impact a small number of developers. Consider that only a fraction of high end developers are defining `ref struct` instances and then consider that only a fraction of those developers will be using these new lifetime features.  That doesn't seem to justify adding a new contextual keyword to the language.

Reviewer 6: Attributes though do seem to escape the language. In this case, will these attributes be meaningful in other .NET languages like VB.NET or F#? If they are applied will they be nops or having some compiler indicate they are unnecessary?

Author: By default these attributes won't be meaningful to other languages. But that's true today. Consider that if they put the `[Dynamic]` attribute on method parameters it has no meaning in F# but will when it's invoked from C#. 

There is a growing list of APIs for which they seem to only have meaning in C#. I'm not sure how or if we should dela with that.

Reviewer 6: > I'm not sure how or if we should dela with that.

Okay. I'll think this does come with some concerns though. From the language it may not matter, but how the runtime may or may not respond to that attribute in the future seems to me to be a potential concern. In fact I could see a strong argument for having tests in those languages to ensure the runtime doesn't do things the generated IL doesn't expect. I don't have a concrete example but I could imagine method level attributes that let the runtime inform the JIT about some lifetime assumption that isn't there in a non-C# language and that ends up causing issues.

I'm in the middle of a GC Stress bug and it is consuming a fair amount of mental resources so it could be simple top of mind and I see potential issues everywhere.

A `ref` field by itself has no impact on span safety rules. As a conecrete example consider that flipping the existing `Span<T>` defintion to use a `ref` field has no impact on consumption. Hence the `ref` itself can be omitted safely. However a `ref` field does have other impacts to consumption that must be preserved: 

- A `ref struct` which has a `ref` field is never considered `unmanaged` 
- The type of the `ref` field impacts infinite generic expansion rules. Hence if the type of a `ref` field contains a type parameter that must be preserved 

Given those rules here is a valid reference assembly transformation for a `ref struct`: 

```c#
// Impl assembly 
ref struct S<T> {
    ref T _field;
}

// Ref assembly 
ref struct S<T> {

>     object _o; // force unmanaged 

Reviewer 4: Was this meant to be either `force not unmanaged` or `force managed`?

Author: yep that is a typo
